plugins/touchegg/Main.vala

```diff
@@ -0,0 +1,159 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+public class Gala.Plugins.Touchegg.Plugin : Gala.Plugin {
+    private Gala.WindowManager? wm = null;
+    private Client? client = null;
+    private GLib.Settings gala_settings;
+    private GLib.Settings touchpad_settings;
+
+    /**
+     * Percentage of the animation to be completed to apply the action.
+     */
+    private const int SUCCESS_THRESHOLD = 20;
+
+    public override void initialize (Gala.WindowManager window_manager) {
+        wm = window_manager;
+        gala_settings = new GLib.Settings ("io.elementary.desktop.wm.gestures");
+        touchpad_settings = new GLib.Settings ("org.gnome.desktop.peripherals.touchpad");
+
+        client = new Client ();
+        client.on_gesture_begin.connect ((gesture) => Idle.add (() => {
+            on_handle_gesture (gesture, "begin");
+            return false;
+        }));
+        client.on_gesture_update.connect ((gesture) => Idle.add (() => {
+            on_handle_gesture (gesture, "update");
+            return false;
+        }));
+        client.on_gesture_end.connect ((gesture) => Idle.add (() => {
+            on_handle_gesture (gesture, "end");
+            return false;
+        }));
+
+        try {
+            client.run ();
+        } catch (Error e) {
+            warning ("Error initializing Touchégg client: %s", e.message);
+        }
+    }
+
+    public override void destroy () {
+        if (client != null) {
+            client.stop ();
+        }
+    }
+
+
+    private void on_handle_gesture (Gesture gesture, string event) {
+        // debug (@"Gesture $(event): $(gesture.type) - $(gesture.direction) - $(gesture.fingers) fingers - $(gesture.percentage)% - $(gesture.elapsed_time) - $(gesture.performed_on_device_type)");
+        var hints = build_hints_from_gesture (gesture, event);
+
+        if (is_open_workspace_gesture (gesture)) {
+            wm.workspace_view.open (hints);
+        }
+
+        if (is_close_workspace_gesture (gesture)) {
+            wm.workspace_view.close (hints);
+        }
+
+        if (is_next_desktop_gesture (gesture)) {
+            if (!wm.workspace_view.is_opened ()) {
+                wm.switch_to_next_workspace (Meta.MotionDirection.RIGHT, hints);
+            }
+        }
+
+        if (is_previous_desktop_gesture (gesture)) {
```

Comment:
I think we could make all the if statements an `else if` here, removing the need to check for other gestures when we already executed an action based on it.